### ISSUE TITLE
Random GAN Training Sets

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#598]
Our Lode Runner results had GANs trained on the first 5 levels, first 20,  first 50 etc. However, what would the GAN output be like if trained on 5 or 20 random levels? Would we get more diversity? Worth trying.

**Comment by schrum2** [comment]
This issue is directly related to work that @kdsteckel did last summer. He generated GANs based on certain level subsets and compared their performance. We then drew some conclusions based on the size of the training set. However, it is a legitimate question to ask if the GAN performance depends more on training set size or composition (or some complex interplay of the two). 

I feel as though Kirby may already have trained many GANs worth exploring. See those in 
https://github.com/schrum2/MM-NEAT/tree/master/src/main/python/GAN/LodeRunnerGAN
@kdsteckel : It looks like you trained a GAN for every group of 10 levels. Can you confirm?

If you run the `LodeRunnerGANLevelBreederTask` class and select the `SetGANModel` button you can load different GAN models and explore their output a bit, albeit informally.

Basically, some fact-finding needs to be done on this issue before going any further.

**Comment by kdsteckel** [comment]
It looks like I trained GANs last summer on 10 level increments, as well as 10 percent increments (15 levels) and 20 percent incemrnents (30 levels). In the LodeRunnerGAN directory.

**Comment by MBatt1** [comment]
for the `LodeRunnerGANModel` parameters, those would correspond to the .pth files in `MM-NEAT\src\main\python\GAN\LodeRunnerGAN`, correct?

**Comment by schrum2** [comment]
I believe that should work. Give it a try and see what happens. It takes a long time to run though. Assuming it doesn't crash immediately, you'll probably want to leave it running over night to see what happens. There will be a new result directory with an archive filled with images of evolved levels.

**Comment by kdsteckel** [comment]
@MBatt1 that should work

**Comment by MBatt1** [comment]
New batch files are as follows:

- Training sets for every 10-level interval; ex 1-10, 11-20, 21-30, ... named like this:
LodeRunnerGAN-MAPElitesGroundTreasureEnemiesCAFit-**OnLevel1to10**
- When run, files output to "loderunnermapelites10batches" folder

- Training sets for every 15-level interval (10%) named like this:
LodeRunnerGAN-MAPElitesGroundTreasureEnemiesCAFit-**On1stTenPercent**
- When run, files output to "loderunnermapelitestenpercentbatches" folder

- Training sets for every 15-level interval (10%) named like this:
LodeRunnerGAN-MAPElitesGroundTreasureEnemiesCAFit-**On1stTwentyPercent**
- When run, files output to "loderunnermapelitestwentypercentbatches" folder

**Comment by schrum2** [comment]
I'm going to re-open this, because I would actually like us to complete at least one training session with each of the sets trained on groups of 10 levels. If the results are promising, then I'd actually like to run more experiments, but this is just some preliminary tests. I'm wondering if some additional results here could improve @kdsteckel 's previous research to the point where it is accepted as a full paper.

I'm going to start running some preliminary experiments myself over the weekend.

**Comment by schrum2** [comment]
![image](https://user-images.githubusercontent.com/8397236/119593003-171a9580-bd9f-11eb-9e9c-49d626c9b0ca.png)
@kdsteckel @MBatt1 Here are results from the Lode Runner runs. They finished a while ago, but it took me some time to create this figure. Basically, I don't think the results are amazing, so I guess we'll drop this line of research and close this issue. However, I'm going to elaborate on what this shows in case it sparks any ideas.

Basically, it looks like the specific set of 10 levels used to train the GAN really do matter, which really isn't surprising in retrospect. Each column is from a different GAN trained in a set of 10 levels identified at the top of the column. 41to50 and 51to 60 seem to produce the most beatable levels, but levels from GANs trained on later levels are generally not beatable. 71to80 fills up the space of possible levels with lots of unbeatable levels, as does 141to150. Ultimately, there is some interesting information here, but it sort of contradicts the hypothesis of our earlier research that training on fewer levels was the deciding factor. 

However, I think this does emphasize the fact that finding beatable Lode Runner levels is hard! So, if we could more reliably find beatable levels somehow, that would be a major accomplishment. Now, I believe that most of these levels genuinely are unbeatable, but I haven't looked closely. It may simply be the case that we need to allow a larger A* budget, but I don't think it's that simple ... evolution could search for levels with shorter solution paths if it were that easy.

So, I think that @MBatt1 might want to focus some effort on Lode Runner if we can ever feel comfortable that CMA-ME #622 is really working. Then again, simply running CMA-ME on Lode Runner and getting good results would be pretty convincing.
